File: src/main/java/ting/leetcode/java/arrays101/BagOfTokens.java
package ting.leetcode.java.arrays101;

import java.util.Arrays;

public class BagOfTokens {
    public int bagOfTokensScore(int[] tokens, int power) {
        if (tokens.length == 0) {
            return 0;
        }

        Arrays.sort(tokens);
        int score = 0;
        int left = 0;
        int right = tokens.length - 1;

        while (left < right) {
            if (tokens[left] <= power) {
                power -= tokens[left];
                score++;
                left++;
            } else if (tokens[right] > power && score > 0) {
                score--;
                power += tokens[right];
                right--;
            } else {
                break;
            }
        }

        if (power >= tokens[right]) {
            score++;
        }

        return score;
    }
}

File: kotlin-solution/src/leetcode/arrays101/BagOfTokens.kt
package leetcode.arrays101

class BagOfTokens {
    fun bagOfTokensScore(tokens: IntArray, power: Int): Int {
        if (tokens.isEmpty()) {
            return 0;
        }

        tokens.sort()
        var score = 0
        var left = 0
        var right = tokens.lastIndex
        var power = power

        while (left < right) {
            if (power >= tokens[left]) {
                power -= tokens[left++]
                score++
            }
            else if (power < tokens[right] && score > 0) {
                power += tokens[right--]
                score--
            }
            else break
        }

        return if (power >= tokens[right]) score + 1 else score
    }
}
